### PR TITLE
document pmap_kextract/vtophys

### PR DESCRIPTION
**Pull request by igalic**
The main body of this pull request is documenting pmap_kextract(9) function and its vtophys(9) alias.

The need to understand pmap_kextract(9) arose from trying to understand virtio code which I am trying to document, and as such this a small side-step in my cloud-init project, where lately I have been working on a virtio vsock driver.

Sponsored by: The FreeBSD Foundation

### CODE REVIEW

**Comment by igalic** [comment]
cc @pauamma who previously reviewed this on Phabricator, before i broke and abandoned those revisions

File: share/man/man9/pmap_kextract.9 [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+function will return the physical page address associated with the
+kernel virtual address
+.Fa va
+inside the physical map of the address space.
+In case of failure,
+.Fn pmap_kextract
+will return
+.Dv NULL .
+.Sh SEE ALSO
+.Xr mutex 9 ,
```
Comment (pauamma) [inline share/man/man9/pmap_kextract.9:47]: Why is this needed?

File: sys/arm/arm/pmap-v6.c [hunk 2]
```diff
@@ -1090,6 +1090,8 @@ pmap_kextract(vm_offset_t va)
 	pt1_entry_t pte1;
```
Comment (igalic) [inline sys/arm/arm/pmap-v6.c:1090]: i should probably copy the comment above this function definition to all implementations.

**Review by igalic — COMMENTED** [review]
another possible improvement.

**Review by pauamma — APPROVED** [review]
Other maybe than the mutex reference, this LGTM and seems to match the intent of the other 3 changes. Consistency with actual source code is beyond me.

File: share/man/man9/pmap_kextract.9 (hunk 1 above)
Comment (igalic) [reply share/man/man9/pmap_kextract.9:47]: that looks like a left-over from copying pmap_extract.9

File: share/man/man9/pmap.9
Comment (grahamperrin) [inline share/man/man9/pmap.9:1]: Need to update the `.Dd` at [`87805c5` (#827)](https://github.com/freebsd/freebsd-src/pull/827/commits/87805c5d374ec884d509eda382cd71ec3b2bd5e8#diff-9eeb8f3f0acc15f299de1876624dc36eca4e0778280f6e901828f6a021b0b773R26).

**Review by grahamperrin — CHANGES_REQUESTED** [review]

File: share/man/man9/pmap_kextract.9 [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+.Fa "vm_offset_t va"
+.Fc
+.Ft vm_paddr_t
+.Fo vtophys
+.Fa "vm_offset_t va"
+.Fc
+.Sh DESCRIPTION
+The
+.Fn pmap_kextract
+function and its alias
```
Comment (grahamperrin) [inline share/man/man9/pmap_kextract.9:31]:
```suggestion
function, and its alias
```

File: share/man/man9/pmap_kextract.9 [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+.Fn vtophys
```
Comment (grahamperrin) [inline share/man/man9/pmap_kextract.9:32]:
```suggestion
.Fn vtophys ,
```

**Review by grahamperrin — CHANGES_REQUESTED** [review]
Maybe add commas. 

I'm not sure that its necessary to mention the alias here (it's already a `.Nm` under _NAME_).

File: share/man/man9/pmap_kextract.9 [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+map a kernel virtual address to a physical page.
+.Pp
+.Sh RETURN VALUES
```
Comment (grahamperrin) [inline share/man/man9/pmap_kextract.9:35]:
```suggestion
.Sh RETURN VALUES
```

Resolve part of <https://github.com/freebsd/freebsd-src/pull/827#issuecomment-1690346027>.

----

From <https://man.freebsd.org/cgi/man.cgi?query=mdoc&sektion=7&manpath=freebsd-release#MACRO_REFERENCE>: 

> … Paragraph breaks are not needed before or after Sh or Ss macros …

**Review by grahamperrin — CHANGES_REQUESTED** [review]

File: share/man/man9/pmap_kextract.9 [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+function will return the physical page address associated with the
+kernel virtual address
+.Fa va
+inside the physical map of the address space.
+In case of failure,
+.Fn pmap_kextract
+will return
+.Dv NULL .
+.Sh SEE ALSO
+.Xr mutex 9 ,
```
Comment (igalic) [inline share/man/man9/pmap_kextract.9:46]:
```suggestion
```

File: share/man/man9/pmap.9
Comment (igalic) [reply share/man/man9/pmap.9:1]: changing the date to today feels overly optimistic.

Comment (grahamperrin) [reply share/man/man9/pmap.9:1]: <https://docs.freebsd.org/en/books/fdp-primer/book/#manual-pages-markup-guidelines> 

> … Trivial changes like spelling or punctuation fixes that do not affect usage …

As the newness of pmap_kextract(9) is significant, so I leant towards treating its `.Xr` in pmap(9) as significant. The cross-reference does not affect usage, however …

… for a person who's accustomed to reading pmap(9) in isolation, a change of date might be a trigger to realise that _something_ significant has changed. 

----

I have minimal experience with manual page review, so I'll leave this for you and other reviewers to decide. 

Thanks

File: sys/amd64/amd64/pmap.c [hunk 7]
```diff
@@ -3933,12 +3933,20 @@ pmap_extract_and_hold(pmap_t pmap, vm_offset_t va, vm_prot_t prot)
...
+ *		Extract from the kernel page table the physical address
+ *		that is mapped by the given virtual address "va".
+ */
 vm_paddr_t
 pmap_kextract(vm_offset_t va)
 {
 	pd_entry_t pde;
 	vm_paddr_t pa;
 
+	pa = 0;
```
Comment (jrtc27) [inline sys/amd64/amd64/pmap.c:3948]: This is dead, pa is always written to below

File: share/man/man9/pmap_kextract.9 [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+The
+.Fn pmap_kextract
+function will return the physical page address associated with the
+kernel virtual address
+.Fa va
+inside the physical map of the address space.
+In case of failure,
+.Fn pmap_kextract
+will return
+.Dv NULL .
```
Comment (jrtc27) [inline share/man/man9/pmap_kextract.9:44]: NULL is of type `void *` but it returns vm_paddr_t, an integer type

File: share/man/man9/pmap_kextract.9 [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+.Sh RETURN VALUES
+The
+.Fn pmap_kextract
+function will return the physical page address associated with the
+kernel virtual address
+.Fa va
+inside the physical map of the address space.
+In case of failure,
+.Fn pmap_kextract
+will return
```
Comment (jrtc27) [inline share/man/man9/pmap_kextract.9:43]: I believe the current contract is "you must be sure it's valid', so panicking is a legitimate implementation, and more helpful perhaps than returning 0 (do we allow 0 as a legitimate physical address?..)

File: sys/arm/arm/pmap-v6.c [hunk 10]
```diff
@@ -1107,9 +1109,6 @@ pmap_kextract(vm_offset_t va)
 		pte2 = pte2_load(pt2map_entry(va));
 		pa = pte2_pa(pte2) | (va & PTE2_OFFSET);
 	}
-	else {
```
Comment (jrtc27) [inline sys/arm/arm/pmap-v6.c:1110]: Stylistically, else pa = 0; is clearer

**Review by jrtc27 — CHANGES_REQUESTED** [review]
If you want to push for this, it needs to be in Phabricator and reviewed by the likes of kib/markj/alc

File: share/man/man9/pmap_kextract.9 [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+The
+.Fn pmap_kextract
+function, and its alias
+.Fn vtophys ,
+map a kernel virtual address to a physical page.
+.Sh RETURN VALUES
+The
+.Fn pmap_kextract
+function will return the physical page address
+.Ft ( vm_paddr_t )
```
Comment (yuripv) [inline share/man/man9/pmap_kextract.9:38]: Use `.Vt` for variable types, i.e. `.Pq Vt vm_paddr_t`.

File: share/man/man9/pmap_kextract.9 [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+associated with the kernel virtual address
+.Fa va
+inside the physical map of the address space.
+In case of failure,
```
Comment (yuripv) [inline share/man/man9/pmap_kextract.9:42]: Please take a look at how this is done in pmap_extract.9 that you are referencing -- there should be a `.Sh RETURN VALUES` section documenting what is returned on success and failure (e.g. address and NULL).  Also I don't think saying it will panic (or fail assertion) makes sense, it shouldn't do that and looking at the code it does not for bad value issues (something being unimplemented or missing from hardware is out of scope of this man page).

**Review by yuripv — CHANGES_REQUESTED** [review]

File: share/man/man9/pmap_kextract.9 [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+.Sh DESCRIPTION
+The
+.Fn pmap_kextract
+function, and its alias
+.Fn vtophys ,
+map a kernel virtual address to a physical page.
+.Sh RETURN VALUES
+The
+.Fn pmap_kextract
+function will return the physical page address
```
Comment (mhorne) [inline share/man/man9/pmap_kextract.9:37]:
```suggestion
function will return the physical address
```

File: share/man/man9/pmap_kextract.9 [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+.Ft ( vm_paddr_t )
+associated with the kernel virtual address
+.Fa va
+inside the physical map of the address space.
```
Comment (mhorne) [inline share/man/man9/pmap_kextract.9:41]:
```suggestion
.Fa va .
```

File: share/man/man9/pmap_kextract.9 (hunk 12 above)
Comment (mhorne) [reply share/man/man9/pmap_kextract.9:42]: I agree here, it would be enough to say that zero/NULL is returned in the case of failure.

Better yet, we should state explicitly in the description the requirement that `va` **must** represent a "valid" kernel address. As in, the value of `va` is obtained by legitimate means (e.g. a pointer value from a malloc'd object). To pass an arbitrary/random value within the KVA range _might_ appear to succeed (if the virtual mapping exists), but it is a programming error subject to "undefined behavior" such as a kernel panic.

I've included this in my suggested text above.

File: share/man/man9/pmap_kextract.9 [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+.\" This manual page was written by Mina Galić <FreeBSD@igalic.co> under
+.\" sponsorship from the FreeBSD Foundation.
+.\"
+.Dd August 24, 2023
+.Dt PMAP_KEXTRACT 9
+.Os
+.Sh NAME
+.Nm pmap_kextract ,
+.Nm vtophys
+.Nd extract from kernel page table to a physical address
```
Comment (mhorne) [inline share/man/man9/pmap_kextract.9:15]:
```suggestion
.Nd extract a physical address from the kernel page table
```

**Review by mhorne — CHANGES_REQUESTED** [review]

File: share/man/man9/pmap_kextract.9 (hunk 11 above)
Comment (igalic) [reply share/man/man9/pmap_kextract.9:38]:
```suggestion
.Pq Vt vm_paddr_t
```

File: share/man/man9/pmap_kextract.9 (hunk 12 above)
Comment (igalic) [reply share/man/man9/pmap_kextract.9]: Thank you for your review.
I've incorporated your changes, and updated this section, too.
If you're still unhappy with it, we can scrap the entire In case of failure paragraph.

File: share/man/man9/pmap_kextract.9 [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+.Ft vm_paddr_t
+.Fo vtophys
+.Fa "vm_offset_t va"
+.Fc
+.Sh DESCRIPTION
+The
+.Fn pmap_kextract
+function, and its alias
+.Fn vtophys ,
+map a kernel virtual address to a physical page.
```
[the remark this answers is not in the document]
Comment (igalic) [reply share/man/man9/pmap_kextract.9]: I've committed this with you as Co-Author.

File: share/man/man9/pmap_kextract.9 [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+.Fa va .
+.Pp
+.Fn pmap_kextract
+generally does not fail.
+However, if supplied with an illegitimate value for
+.Fa va
+it will return 0 on arm64, and
+.Xr panic 9
+on armv6 and riscv.
+Note that such platform dependent behaviours are subject to change.
```
Comment (jrtc27) [inline share/man/man9/pmap_kextract.9:56]: What's the point of documenting it? Nobody should be relying on the behaviour. And this is incomplete anyway.

File: share/man/man9/pmap_kextract.9 [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+associated with the kernel virtual address
+.Fa va .
+.Pp
+.Fn pmap_kextract
+generally does not fail.
+However, if supplied with an illegitimate value for
+.Fa va
+it will return 0 on arm64, and
+.Xr panic 9
+on armv6 and riscv.
```
Comment (jrtc27) [inline share/man/man9/pmap_kextract.9:55]: armv6 is a MACHINE_ARCH not a MACHINE (which would be arm) like arm64 and riscv (and armv7 exists)

File: sys/arm64/arm64/pmap.c [hunk 19]
```diff
@@ -1949,6 +1949,10 @@ pmap_klookup(vm_offset_t va, vm_paddr_t *pa)
 	return (true);
 }
 
+/*
+ *  Extract from the kernel page table the physical address
```
Comment (jrtc27) [inline sys/arm64/arm64/pmap.c:1953]: Why the change in style for arm64 and riscv?

File: sys/amd64/amd64/pmap.c [hunk 20]
```diff
@@ -3933,6 +3933,12 @@ pmap_extract_and_hold(pmap_t pmap, vm_offset_t va, vm_prot_t prot)
 	return (m);
 }
 
+/*
+ *	Routine:	pmap_kextract
+ *	Function:
+ *		Extract from the kernel page table the physical address
```
Comment (jrtc27) [inline sys/amd64/amd64/pmap.c:3939]: Can this be made to more closely follow the wording for pmap_extract? Something more like:

> Extract the physical page address associated with the given kernel virtual address.

**Review by mhorne — APPROVED** [review]
Thanks! I can merge this.

**Comment by igalic** [comment]
merged in 74e4a8d208f0b3cf2525e3786f3efba71fcdb752
